File: pattern-matching/kotlin/response.kt
/* Hiérarchie de classes pour illustrer le pattern matching structurel */

sealed class Response<out A: Any>
object Pending: Response<Nothing>()
data class Failure(val message: String): Response<Nothing>()
data class Success<A: Any>(val result: A, val comment: String? = null): Response<A>()

fun<A: Any> processResponse(reponse: Response<A>) = when(reponse) {
    is Pending -> "rien reçu"
    is Failure ->
        if (reponse.message == "tout va bien") "on a reçu un paradoxe!"
        else "on a reçu une erreur: " + reponse.message
    is Success<A> -> reponse.comment?.let {
        "ça a fonctionné et on nous a signalé: $it"
    } ?: "ça a fonctionné!"
}

fun main() {
    for (test in listOf(
        Pending,
        Success<Int>(42),
        Success<Int>(42, "quelle est la question?"),
        Failure("catastrophe"),
        Failure("tout va bien"),
    )) processResponse(test).also(::println)
}

File: pattern-matching/scala/response.scala
/* Hiérarchie de classes pour illustrer le pattern matching structurel */

sealed trait Response[+A]

object Pending extends Response[Nothing]
case class Failure(message: String) extends Response[Nothing]
case class Success[+A](result: A, comment: Option[String] = None) extends Response[A]

object Main extends App {

  def processResult[A](response: Response[A]): String =
    response match {
      case Pending => "rien reçu"
      case Failure(message) if message == "tout va bien" => "on a reçu un paradoxe!"
      case Failure(message) => "on a reçu une erreur: " + message
      case Success(result, None) => "ça a fonctionné!"
      case Success(result, Some(comment))
          => "ça a fonctionné et on nous rapporte: " + comment
    }

  for (test <- List(
    Pending,
    Success[Int](42),
    Success[Int](42, Some("quelle est la question?")),
    Failure("catastrophe"),
    Failure("tout va bien"),
  )) {
    println(processResult(test))
  }
}
